cpp/src/replace/clamp.cu

```diff
@@ -0,0 +1,168 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/types.hpp>
+#include <cudf/scalar/scalar.hpp>
+#include <cudf/detail/copy.hpp>
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/replace.hpp>
+#include <cudf/detail/iterator.cuh>
+
+namespace cudf {
+namespace experimental {
+namespace detail {
+namespace {
+struct dispatch_clamp {
+    template <typename T, typename Predicate>
+    void apply_transform (column_device_view  input,
+                          mutable_column_device_view output,
+                          Predicate pred,
+                          cudaStream_t stream)
+    {
+        if (input.nullable()){
+            auto input_begin = cudf::experimental::detail::make_null_replacement_iterator<T>(input);
+            thrust::transform(rmm::exec_policy(stream)->on(stream),
+                              input_begin,
+                              input_begin+input.size(),
+                              detail::make_validity_iterator(input),
+                              output.begin<T>(),
+                              pred);
+        } else {
+            thrust::transform(rmm::exec_policy(stream)->on(stream),
+                              input.begin<T>(),
+                              input.end<T>(),
+                              output.begin<T>(),
+                              pred);
+        }
+    }
+
+    template <typename T>
+    std::enable_if_t<cudf::is_fixed_width<T>(), std::unique_ptr<cudf::column>>
+    operator()(column_view const& input,
+              scalar const& lo,
+              scalar const& hi,
+              rmm::mr::device_memory_resource* mr,
+              cudaStream_t stream) {
+        using ScalarType = cudf::experimental::scalar_type_t<T>;
+        auto lo_scalar = static_cast<ScalarType const*>(&lo);
+        auto hi_scalar = static_cast<ScalarType const*>(&hi);
+        auto lo_value = lo_scalar->value(stream);
+        auto hi_value = hi_scalar->value(stream);
+        // mask will not change
+        std::unique_ptr<column> output {nullptr};
+        if (input.nullable()){
+            output = make_fixed_width_column(input.type(), input.size(),
+                                             copy_bitmask(input, stream, mr),
+                                             input.null_count(), stream, mr);
+        } else {
+            output = allocate_like(input, input.size(), mask_allocation_policy::NEVER, mr, stream);
+        }
+
+        auto output_device_view  = cudf::mutable_column_device_view::create(output->mutable_view(), stream);
+        auto input_device_view  = cudf::column_device_view::create(input, stream);
```

Comment:
Consider using `make_pair_iterator` for the input. Then the input to `apply_transform` can just be an iterator instead of a `column_device_view`.

Reply:
Do we have `make_pair_iterator`, I don't see it for libcudf++

Reply:
Waiting on #3490